conn.go

```diff
@@ -1347,7 +1527,8 @@ func (c *Conn) executeQuery(ctx context.Context, qry *Query) *Iter {
 		params.pageSize = qry.pageSize
 	}
 	if c.version > protoVersion4 {
-		params.keyspace = c.currentKeyspace
+		params.keyspace = qry.keyspace
```

Comment:
I just realized that we're still using `c.currentkeyspace` in the prepared statement cache key, if the user provides a different keyspace then we have to use that one. This is happening in `executeQuery`, `executeBatch`, `prepareStatementForKeyspace`. These are the ones I managed to catch. Idk if it is easy to write a test for this

Reply:
You can probably catch this bug if you use the same query string on 2 statements but one of them you call `SetKeyspace`, if I'm correct both statements will internally use the same prepared statement and will go towards the same keyspace.

Reply:
This is a very good catch. The driver should use the bounded keyspace of the stmt. At this moment, the key contains `c.currentKeyspace` which is the initial keyspace specified at `Cluster.Keyspace` and it looks like this:
```
Cluster.Keyspace = "__INITIAL_KS__"

98267857-1640-4054-92a1-9d214745ba9e__INITIAL_KS__SELECT * FROM gocql_test.keyspace_bug
```

I'm proposing to change this a bit to use keyspace in a cache key only when the QUERY/BATCH has an overridden keyspace (e.g. `SetKeyspace`).

So, if the query has an overridden keyspace the stmt cache key would look like this:
```
session.Query("SELECT * FROM gocql_test.keyspace_bug").SetKeyspace("__OVERRIDEN_KEYSPACE__")

98267857-1640-4054-92a1-9d214745ba9e__OVERRIDEN_KEYSPACE__SELECT * FROM gocql_test.keyspace_bug
```

If there is no overridden keyspace for the query, the keyspace won't be presented in the stmt key:
```
session.Query("SELECT * FROM gocql_test.keyspace_bug")

98267857-1640-4054-92a1-9d214745ba9eSELECT * FROM gocql_test.keyspace_bug
```

Also, I found that `routingKeyInfo` also uses `c.current` keyspace and this place is a bit tricky for me. I can just make it to take a keyspace, but I'm not sure how it lines with the purpose of this method...

And one more thing, if we agree on the proposal, we should get rid of `c.currentKeyspace`.

Reply:
> If there is no overridden keyspace for the query, the keyspace won't be presented in the stmt key:

Hmm I wonder what happens if you change the bounded keyspace on the connection and reuse the prepared statement that was prepared before the keyspace change? If C* uses the keyspace at the time of the PREPARE then this will not work (since gocql internally prepares automatically and the user might not even know this is happening).

Reply:
I think we have to use `c.currentkeyspace` in the cache key if the keyspace wasn't set on the request.

Reply:
> if you change the bounded keyspace on the connection

As far as I see gocql doesn't expose an API to change the bounded keyspace for the connection. The only place I see where initialized `Conn` is exposed is `HandleErorrHandler.HandleError` which is not meant to be used this way (?). However, there is still a possibility to call `Conn.UseKeyspace`...

Then we could just use the bounded keyspace if it is present, and when it is not we have to use `c.currentKeyspace`.

Reply:
Hmm yeah it's true that gocql doesn't expose an API and it also tries to prohibit `USE` statements but it's quite easy to "trick" gocql into allowing it by just adding spaces or `;`.

> However, there is still a possibility to call Conn.UseKeyspace.
It's a bit weird to me that the `Conn` type has exported functions like that but that's a conversation for another time 😅

Reply:
Hmm this prompted me to look a bit further into the routing key info and I think there's a couple of places where we are using the keyspace from the prepared metadata without any fallback but there's no guarantee that the metadata will have that keyspace field populated.

[Here](https://github.com/worryg0d/gocql/blob/b5d76bd04c9f32207554dd27f7cb6369423f5365/conn.go#L1589) we can set it to `usedKeyspace` if `info.request.keyspace == ""`

[And here](https://github.com/worryg0d/gocql/blob/b5d76bd04c9f32207554dd27f7cb6369423f5365/session.go#L663) we can set it to `keyspace` if `info.request.keyspace == ""`

On a separate note I think we're also missing a check if keyspace was set on the query [here](https://github.com/worryg0d/gocql/blob/b5d76bd04c9f32207554dd27f7cb6369423f5365/session.go#L1151-L1165), it should take precedence over the one from the `cfg`:

```
	if q.getKeyspace != nil {
		return q.getKeyspace()
	}
	if q.routingInfo.keyspace != "" {
		return q.routingInfo.keyspace
	}
        if q.keyspace != "" {
                return q.keyspace
        }

	if q.session == nil {
		return ""
	}
	// TODO(chbannis): this should be parsed from the query or we should let
	// this be set by users.
	return q.session.cfg.Keyspace
```

Reply:
It's unclear when the `<global_table_spec>` flag is set in response to **PREPARE**. When I was testing this behavior C* always responded with this flag on stmt prepare requests. However, it wouldn't if we double-check this. I added this to the last commit.